File: app/src/main/java/com/example/android_kotlin_handson/RepositoryListFragment.java
package com.example.android_kotlin_handson;

import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.example.android_kotlin_handson.api.RetrofitSearchRepositoryApi;
import com.example.android_kotlin_handson.api.SearchRepositoryApi;

import java.util.ArrayList;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;

public class RepositoryListFragment extends Fragment {
    private static final String ARG_LANGUAGE = "ARG_LANGUAGE";

    private String mLanguage;
    private SearchRepositoryApi api;
    private CompositeDisposable disposables;

    public RepositoryListFragment() {
        this.api = new RetrofitSearchRepositoryApi();
    }

    public static RepositoryListFragment newInstance(String language) {
        RepositoryListFragment fragment = new RepositoryListFragment();
        Bundle args = new Bundle();
        args.putString(ARG_LANGUAGE, language);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            mLanguage = getArguments().getString(ARG_LANGUAGE);
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_repository_list, container, false);

        RecyclerView recyclerView = (RecyclerView) view.findViewById(R.id.recyclerView);
        recyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));
        recyclerView.setAdapter(new RecyclerAdapter(getContext(), new ArrayList<>()));
        return view;
    }

    @Override
    public void onResume() {
        super.onResume();
        disposables = new CompositeDisposable();

        disposables.add(api.searchRepositoryByLanguage(mLanguage)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(repositoryList -> {
                    View v = getView();
                    if (v != null) {
                        RecyclerView recyclerView = v.findViewById(R.id.recyclerView);
                        recyclerView.swapAdapter(new RecyclerAdapter(getContext(), repositoryList), false);
                    }
                }));
    }

    @Override
    public void onStop() {
        super.onStop();
        if (disposables != null) {
            disposables.dispose();
        }
        disposables = null;
    }
}

File: app/src/main/java/com/example/android_kotlin_handson/RepositoryListFragment.kt
package com.example.android_kotlin_handson

import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.example.android_kotlin_handson.api.RetrofitSearchRepositoryApi
import com.example.android_kotlin_handson.api.SearchRepositoryApi
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable

class RepositoryListFragment : Fragment() {

    private val language: String by lazy {
        arguments?.getString(ARG_LANGUAGE) ?: ""
    }

    private val api: SearchRepositoryApi = RetrofitSearchRepositoryApi()

    private var disposables: CompositeDisposable? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_repository_list, container, false)

        view.findViewById<RecyclerView>(R.id.recyclerView).run {
            layoutManager = LinearLayoutManager(activity)
            adapter = RecyclerAdapter(context, listOf())
        }
        return view
    }

    private fun Disposable.into(disposable: CompositeDisposable?) {
        disposable?.add(this)
    }

    override fun onResume() {
        super.onResume()
        disposables = CompositeDisposable()
        api.searchRepositoryByLanguage(language!!)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe { repositoryList ->
                    view?.let {
                        val recyclerView = it.findViewById<RecyclerView>(R.id.recyclerView)
                        recyclerView.swapAdapter(RecyclerAdapter(context!!, repositoryList), false)
                    }}
                .into(disposables)
    }

    override fun onStop() {
        super.onStop()
        disposables?.dispose()
        disposables = null
    }

    companion object {
        private const val ARG_LANGUAGE = "ARG_LANGUAGE"

        fun newInstance(language: String) =
                RepositoryListFragment().apply {
                    arguments = Bundle().apply {
                        putString(ARG_LANGUAGE, language)
                    }
                }
    }
}
